### PR TITLE
logging/ascii: Fix .shadow paths when using LogAscii::logdir

### PR DESCRIPTION
**Pull request by awelzel**
With the introduction of LogAscii::logdir, log filenames can now include
parent directories rather than being plain basenames. Enabling log rotation,
leftover log rotation and setting LogAscii::logdir broke due to not
handling this situation.

This change ensures that .shadow files are placed within the directory where
the respective .log file is created. Previously, the .shadow. (or .tmp.shadow.)
prefix was simply prepended, yielding non-sensical paths such as
.tmp.shadow.foo/bar/packet_filter.log for a logdir of foo/bar.

Additionally, respect LogAscii::logdir when searching for leftover log files
rather than defaulting to the current working directory.

The following quirk exist around LogAscii::logdir, but will be
treated separately.

* By default, logs are currently rotated into the working directory of the
  process, rather than staying confined within LogAscii::logdir. One of
  the added tests shows this behavior.

### CODE REVIEW

File: testing/btest/scripts/base/frameworks/logging/rotate-logdir-leftover-log-rotation-shadow-files.zeek [hunk 1]
```diff
@@ -0,0 +1,29 @@
+# @TEST-DOC: Test that .shadow files are picked up from LogAscii::logdir.
```
Comment (rsmmr) [inline testing/btest/scripts/base/frameworks/logging/rotate-logdir-leftover-log-rotation-shadow-files.zeek:1]: Do we have existing tests that cover the case of logdir not being set?

**Review by rsmmr — COMMENTED** [review]
Nice, this looks good.

I'm just wondering if we should now actually wait for the `filesystem` library to land after all, given that we have consensus (and a PR) on that already; then replace the manual path manipulation with that?

File: src/logging/writers/ascii/Ascii.cc [hunk 2]
```diff
@@ -83,19 +84,88 @@ struct LeftoverLog
...
+	CHECK(prefix_basename_with("", ".shadow.") == ".shadow.");
+	CHECK(prefix_basename_with("conn.log", ".shadow.") == ".shadow.conn.log");
+	CHECK(prefix_basename_with("/conn.log", ".shadow.") == "/.shadow.conn.log");
+	CHECK(prefix_basename_with("a/conn.log", ".shadow.") == "a/.shadow.conn.log");
+	CHECK(prefix_basename_with("/a/conn.log", ".shadow.") == "/a/.shadow.conn.log");
+	CHECK(prefix_basename_with("a/b/conn.log", ".shadow.") == "a/b/.shadow.conn.log");
+	}
+
+/**
+ * Trim trailing slashes from directory path.
```
[the remark this answers is not in the document]
Comment (awelzel) [reply src/logging/writers/ascii/Ascii.cc:140]: Good point. The looping makes sense. But, agree that `filesystem` would help so moving that PR forward first should make this less arcane.

File: src/logging/writers/ascii/Ascii.cc [hunk 3]
```diff
@@ -83,19 +84,88 @@ struct LeftoverLog
...
+	std::string filename = path;
+	if ( last != std::string::npos )
+		{
+		dirname = path.substr(0, last);
+		filename = path.substr(last + 1);
+		}
+
+	filename = prefix + filename;
+
+	if ( dirname.empty() && last == std::string::npos )
```
[the remark this answers is not in the document]
Comment (awelzel) [reply src/logging/writers/ascii/Ascii.cc:123]: Hmm, this was "special casing" an input of `/x.log` being split into an empty `dirname`, but the result should continue to be absolute. Again, hope `filesystem` makes this better.

[the remark this answers is not in the document]
Comment (awelzel) [reply src/logging/writers/ascii/Ascii.cc]: Repushed a version that's using ghc::filesystem for this.

File: src/logging/writers/ascii/Ascii.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (awelzel) [reply src/logging/writers/ascii/Ascii.cc]: This helper is gone, now using `/` provided on `ghc::filesystem::path`

**Comment by awelzel** [comment]
@rsmmr  - I have continued on this, used ghc::filesystem and introduced `Log::default_logdir`. Seems ready for you to take another look.

I may have gotten some details about the deprecation wrong, so please point out anything that seems off.

Thanks!

**Comment by awelzel** [comment]
The changes in this PR address #2225 as well.

**Review by rsmmr — COMMENTED** [review]
Looks good.  I think those deprecations should work.

Just one nit, and please squash that final fixup commit.

**Review by rsmmr — APPROVED** [review]

**Pull request merged**